packages/wallet-sdk/src/core/communicator/Communicator.ts

```diff
@@ -96,5 +110,40 @@ export class Communicator {
         if (!this.popup) throw standardErrors.rpc.internal();
         return this.popup;
       });
-  }
+  };
+
+  /**
+   * Posts a RPC request to the popup window and waits for a response
+   */
+  postRPCRequest = (request: RPCRequestMessage): Promise<RPCResponseMessage> => {
+    this.pendingRequests.set(request.id, 'created');
+
+    return new Promise((resolve, reject) => {
+      // defer the execution the promise until the current call stack clears
+      Promise.resolve().then(() => {
+        // reject if the request was cancelled between creation and awaiting due to PopupUnload
+        if (!this.pendingRequests.has(request.id)) {
+          reject(standardErrors.provider.userRejectedRequest('Request cancelled before sending'));
+          return;
+        }
+
+        // otherwise, post the message and wait for a response
+        this.pendingRequests.set(request.id, 'awaiting');
+        this.postMessage(request)
+          .then(() =>
+            this.onMessage<RPCResponseMessage>(({ requestId }) => requestId === request.id)
+          )
+          // resolve or reject the promise to take any follow-up requests if necessary before disconnecting
+          .then(resolve)
+          .catch(reject)
+          // then clean up the current request and disconnect if there are no more pending requests
+          .finally(() => {
+            this.pendingRequests.delete(request.id);
+            if (this.pendingRequests.size === 0) {
+              this.disconnect();
+            }
```

Comment:
regardless of the result, all `awaiting` requests will eventually hit this block. if there are no pending requests, it can close the popup.